### ISSUE TITLE
[REVIEW]: c-lasso: a Python package for constrained sparse regression and classification

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2844]
**Submitting author:** @Leo-Simpson (<a href="http://orcid.org/0000-0003-0717-5244">Leo Simpson</a>)
**Repository:** <a href="https://github.com/Leo-Simpson/c-lasso" target ="_blank">https://github.com/Leo-Simpson/c-lasso</a>
**Version:** v1.0
**Editor:** @mjsottile
**Reviewer:** @jbytecode, @glemaitre
**Archive:** <a href="https://doi.org/10.6084/m9.figshare.13589585.v1" target="_blank">10.6084/m9.figshare.13589585.v1</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46/status.svg)](https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46"><img src="https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46/status.svg)](https://joss.theoj.org/papers/e830fbd5fa1c1bc5480e73979a992b46)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/Leo-Simpson/c-lasso" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jbytecode & @glemaitre, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mjsottile know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Leo-Simpson/c-lasso">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@Leo-Simpson) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Leo-Simpson/c-lasso">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@Leo-Simpson) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jbytecode, @glemaitre it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/biomet/71.2.323 is OK
- 10.1101/2020.09.01.277632 is OK
- 10.1007/s10957-018-1430-2 is OK
- 10.1007/s11228-011-0191-y is OK
- 10.1214/19-EJS1662 is OK
- 10.1007/s12561-020-09283-2 is OK
- 10.1080/10618600.2018.1473777 is OK
- 10.1080/01621459.2019.1609970 is OK
- 10.1007/bf02985802 is OK
- 10.1016/j.csda.2020.106958 is OK
- 10.1162/NECO_a_00434 is OK
- 10.1093/biomet/asu031 is OK
- 10.1111/j.1467-9868.2010.00740.x is OK
- 10.1214/009053606000001370 is OK
- 10.1214/16-AOAS928 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by jbytecode** [comment]
@whedon check repository

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.84  T=10.07 s (4.3 files/s, 9816.0 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
HTML                             3           6745              6          41729
SQL                              1              0              0          37299
Python                          19            897            821           3455
Markdown                         6            499              3           1113
Jupyter Notebook                 3              0           5152            520
TeX                              1             21              1            210
R                                3             35             32             76
reStructuredText                 5             52             64             37
DOS Batch                        1              8              1             26
make                             1              4              7              9
-------------------------------------------------------------------------------
SUM:                            43           8261           6087          84474
-------------------------------------------------------------------------------


Statistical information for the repository '3ed71301e422b11a6362b526' was
gathered on 2020/11/18.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Leo-Simpson                    168         14728           9555          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Leo-Simpson                5173           35.1          2.9                7.04
```

**Comment by whedon** [comment]
Failed to discover a valid open source license.

**Comment by jbytecode** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by jbytecode** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:wave: @glemaitre, please update us on how your review is going.

**Comment by whedon** [comment]
:wave: @jbytecode, please update us on how your review is going.

**Comment by jbytecode** [comment]
I have opened many issues. The authors have finished some of them. They are also working on the other issues. Whenever an issue is closed, I check the corresponding review item.

**Comment by mjsottile** [comment]
:wave: @glemaitre I wanted to check to see if you had a chance to start your review?  Let me know if you need anything to proceed.

**Comment by glemaitre** [comment]
Thanks for pinging. I will make a full pass now. Sorry for the delay.

**Comment by glemaitre** [comment]
I am adding a question here since that I would like to have advice from @mjsottile 

So one statement in the package is the following:

> his package is intended to fill the gap between popular python tools such as scikit-learn which CANNOT solve sparse constrained problems and general-purpose optimization solvers that do not scale well for the considered problems.

 It is completely correct that scikit-learn does not provide this functionality. So, one of my expectation would be to be able to use `c-lasso` as a drop-in replacement instead of scikit-learn to solve such problem. However, I might be bias by being involve in the development of scikit-learn. The issue here is that you cannot use the package for such purpose due to API issue incompatibilities.

Indeed, it would require a tremendous amount of work but this is possible. I can link to the following package which shows such of an example (https://mathurinm.github.io/celer/index.html). I would expect Python functions to implement the different solvers and a common class with parameters allowing to select a specific solver. Turning on/off flags is not really something done in scikit-learn. With such design, `c-lasso` would be used by people from the scikit-learn community while currently it would not be possible.

I see that @Leo-Simpson is also using the package with R which is probably one of the reason of having different vison regarding the package.

@mjsottile Indeed, I would like to know if these thoughts should be considered while reviewing the paper. I would not like to block the review/acceptance process on a criterion that should not be considered in the review process.

**Comment by muellsen** [comment]
Thanks for raising this important issue. Indeed, our initial hope was to design c-lasso as drop-in replacement in scikit-learn and use all the cool functionality there. However, after an in-depth analysis at the start of the project in early 2019, we opted for a stand-alone version. Below is our thinking and a note about potential future developments 
- First, we do not have yet the resources and knowledge to do such an integration. I spent several days understanding the implementations of the optimization schemes underlying the standard lasso problems and found them to be incompatible with our ideas (largely due to my inadequacies I think)    
- Second, c-lasso is indeed integral for several R packages (including one already available, called "trac") that we are currently developing, and thus, the current solution makes it easier (for us) to handle these joint developments
- Third, c-lasso is integral for future QIIME-2  plugins (developed in parallel) where we again would love to have light-weight dependencies and full control over the solver development (this is specific to the needs in the microbiome data analysis community)
- Forth, if someone from the scikit-learn community is interested in this, we would love to collaborate. I know that Alexandre Gramfort and colleagues have worked on similar ideas regarding concomitant estimation (so problems with estimating sigma) so maybe there is a future connection?!

**Comment by mjsottile** [comment]
Hi @glemaitre - good questions, and thanks for the details @muellsen.  I think this issue is worth addressing in the paper explicitly clarify that, in its current state, the package is not an API-compatible drop in replacement for components of scikit-learn.  The issue with API compatibility (or the lack thereof) doesn't detract from the quality of the work itself, just the expectations that users may have.  If the paper just adds a simple statement that the code currently is not a straight drop-in replacement in scikit-learn, then potential users won't mistakenly assume that it is.  I think such a design for c-lasso would be interesting to look at in the future, but isn't critical for this submission.

**Comment by glemaitre** [comment]
OK, so this is fine with me. I just open a couple of issues such that the content of the README is included in the documentation webpage and to have a second look regarding the test coverage.

Now I am hooked so I am might propose a bit of help regarding the scikit-learn API part :)

**Comment by mjsottile** [comment]
:wave: @glemaitre and @jbytecode : I've been tracking progress in the issues linked above, but in the interest of being able to respond in a timely manner once reviews are complete, please add a comment to this issue thread when you believe you are finished and indicate whether or not you support accepting the submission (and if not, what you believe must be addressed). 

Please let me know if you have any questions/concerns as well.  Thanks, and Happy New Year!  :partying_face:

**Comment by jbytecode** [comment]
Dear editor @mjsottile,

In my reviewing process, I opened many issues indicating the problems by referencing the corresponding reviewing items, and the authors corrected them. I generally respond with a message "that seems okay" before closing the issue. Please tell me if there is a specific problem with this and let me correct. 

A single issue is waiting for correction, the authors wrote a micro benchmark tool and generated some graphics. In the [issue](https://github.com/Leo-Simpson/c-lasso/issues/6) they reported that they would present the results in the paper. I am currently waiting them to add these reports.

Happy new year! Thank you very much for your great efforts in JOSS!

**Comment by muellsen** [comment]
> Hi @glemaitre - good questions, and thanks for the details @muellsen. I think this issue is worth addressing in the paper to explicitly clarify that, in its current state, the package is not an API-compatible drop in replacement for components of scikit-learn. The issue with API compatibility (or the lack thereof) doesn't detract from the quality of the work itself, just the expectations that users may have. If the paper just adds a simple statement that the code currently is not a straight drop-in replacement in scikit-learn, then potential users won't mistakenly assume that it is. I think such a design for c-lasso would be interesting to look at in the future, but isn't critical for this submission.

@glemaitre @mjsottile I added a line in the Paper.md and the README clarifying that c-lasso is a stand-alone package and not yet compatible with the scikit-learn API. It would be absolutely awesome, however, to work on this!

**Comment by muellsen** [comment]
> Dear editor @mjsottile,
> 
> In my reviewing process, I opened many issues indicating the problems by referencing the corresponding reviewing items, and the authors corrected them. I generally respond with a message "that seems okay" before closing the issue. Please tell me if there is a specific problem with this and let me correct.
> 
> A single issue is waiting for correction, the authors wrote a micro benchmark tool and generated some graphics. In the [issue](https://github.com/Leo-Simpson/c-lasso/issues/6) they reported that they would present the results in the paper. I am currently waiting them to add these reports.
> 
> Happy new year! Thank you very much for your great efforts in JOSS!

We added the relevant benchmarks to the GitHub and the Paper

**Comment by jbytecode** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by jbytecode** [comment]
@muellsen, Please check the pdf

- L128: `1,8` might be `1.8` ?
- L129: `8 Go` might be `8 Gb`?

**Comment by jbytecode** [comment]
Dear authors,

please update your status. 

thanks. @muellsen

**Comment by muellsen** [comment]
> @muellsen, @Leo-Simpson, Please check the pdf
> 
> * L128: `1,8` might be `1.8` ?
> * L129: `8 Go` might be `8 Gb`?

Just saw this, sorry. Yes. Updated!

**Comment by muellsen** [comment]
Both in .md file and benchmark folder

**Comment by jbytecode** [comment]
Dear editor @mjsottile 

- The authors added the missing license entry in the software repository.
- Unit test were missing. The authors added automation tests. Now, the package is automatically tested before being available. 
- The paper is well written in my opinion. Beyond this, they corrected all of the typos and code formatting that I addressed.
- There were serious performance claims in the earlier versions of the paper. The authors added some benchmark tools to software repository and they extended the paper with a new section that includes reports of these benchmark studies. I think that is okay.
- The authors also added community guidelines.

Summarizing all, the authors considered all of the issues that I addressed. Personally, I can recommend an acceptance.

**Comment by jbytecode** [comment]
> > @muellsen, @Leo-Simpson, Please check the pdf
> > 
> > * L128: `1,8` might be `1.8` ?
> > * L129: `8 Go` might be `8 Gb`?
> 
> Just saw this, sorry. Yes. Updated!

Thank you @muellsen, I am done with this. My review report is above.

**Comment by muellsen** [comment]
@jbytecode Thank you very very much. Much appreciated that you dedicated your time to this! This is awesome!

**Comment by mjsottile** [comment]
@whedon generate pdf

**Comment by mjsottile** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/biomet/71.2.323 is OK
- 10.1101/2020.09.01.277632 is OK
- 10.1007/s10957-018-1430-2 is OK
- 10.1007/s11228-011-0191-y is OK
- 10.1214/19-EJS1662 is OK
- 10.1007/s12561-020-09283-2 is OK
- 10.1080/10618600.2018.1473777 is OK
- 10.1080/01621459.2019.1609970 is OK
- 10.1007/bf02985802 is OK
- 10.1016/j.csda.2020.106958 is OK
- 10.1162/NECO_a_00434 is OK
- 10.1093/biomet/asu031 is OK
- 10.1111/j.1467-9868.2010.00740.x is OK
- 10.1214/009053606000001370 is OK
- 10.1214/16-AOAS928 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by mjsottile** [comment]
Hi @muellsen @Leo-Simpson - I am making my final editorial pass over the text.  A couple questions about the text:

- I noticed a number of places where the word "constrained" is spelled "contrained".  I am assuming this is a typo that you can search and replace to fix.
- In the toy example, it would be useful to reformat slightly so that the closing paren for the random_data() call isn't on its own at the top of the following page.
- Remove capitalization of "laptop" at the end of the section on the toy example.  
- Capitalize "Lipschitzian" in the references since it derives from a proper noun.

Once those edits are complete, I believe there will be no remaining outstanding issues and we can proceed to finalizing the paper.  The final steps will be:

- Submit a version of the software to an archive (e.g., figshare or zenodo) and post the DOI for the archive in this thread.

- Please verify that the version number matches that of the archived code, and if not, let me know so I can update the version here.

- Finally, make sure that the metadata for the archive (e.g., title, authors) matches the paper.

Please comment here when you've completed everything and have a DOI.

**Comment by Leo-Simpson** [comment]
Dear @mjsottile, 

Thank you very much. 

I finished those edits, except for the problem in the references, which I think can not be fixed on the paper, because the references are generated automatically with pandoc with a certain citation rendering, and I assume that it restricts the rendering of the title, because there is a capital letter in the `paper.bib`file. 

I submitted the software to figshare : https://doi.org/10.6084/m9.figshare.13589585.v1

So I guess the version of the archived code is 'v1', but I don't know if we can modify it. 

The DOI is : 10.6084/m9.figshare.13589585 

Best regards.

**Comment by mjsottile** [comment]
@Leo-Simpson I believe you can use the usual BiBTeX method to preserve case by wrapping the titles (or individual words) in an additional layer of {} braces.  Pandoc is just using LaTeX to render the markdown into PDF, so that should fix the cases.  It looks like some of your bib file entries already have extra braces like this (eg, see the title entry for bien:2020).

**Comment by Leo-Simpson** [comment]
@mjsottile Thank you for the indication, I did not know how to do this. I have just edited it.

**Comment by mjsottile** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02844/joss.02844/10.21105.joss.02844.pdf) :page_facing_up: :point_left:

**Comment by mjsottile** [comment]
@whedon set https://doi.org/10.6084/m9.figshare.13589585.v1 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.6084/m9.figshare.13589585.v1" target="_blank">10.6084/m9.figshare.13589585.v1</a> is the archive.

**Comment by mjsottile** [comment]
@whedon set v1.0 as version

**Comment by whedon** [comment]
OK. v1.0 is the version.

**Comment by mjsottile** [comment]
Everything looks good to me @Leo-Simpson - I will hand the paper to the EICs for final processing.

Thank you very much @glemaitre and @jbytecode for your time and thoughtful reviews.

**Comment by mjsottile** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/biomet/71.2.323 is OK
- 10.1101/2020.09.01.277632 is OK
- 10.1007/s10957-018-1430-2 is OK
- 10.1007/s11228-011-0191-y is OK
- 10.1214/19-EJS1662 is OK
- 10.1007/s12561-020-09283-2 is OK
- 10.1080/10618600.2018.1473777 is OK
- 10.1080/01621459.2019.1609970 is OK
- 10.1007/bf02985802 is OK
- 10.1016/j.csda.2020.106958 is OK
- 10.1162/NECO_a_00434 is OK
- 10.1093/biomet/asu031 is OK
- 10.1111/j.1467-9868.2010.00740.x is OK
- 10.1214/009053606000001370 is OK
- 10.1214/16-AOAS928 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2028

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2028, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1350732626500456448) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2029
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02844](https://doi.org/10.21105/joss.02844)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@jbytecode, @glemaitre - many thanks for your reviews here and to @mjsottile for editing this submission. JOSS relies upon the volunteer efforts of folks like yourselves and we simply wouldn't be able to do this without you! ✨ 

@Leo-Simpson - your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02844/status.svg)](https://doi.org/10.21105/joss.02844)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02844">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02844/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02844/status.svg
   :target: https://doi.org/10.21105/joss.02844
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02844/status.svg)](https://doi.org/10.21105/joss.02844)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
